File: kakao.app/js/kxk/test/kseg.js
var toExport = {}
var hello, text

import kstr from "../kstr.js"
import kseg from "../kseg.js"

toExport["kseg"] = function ()
{
    hello = `hello
world`
    section("kseg", function ()
    {
        compare(kseg(hello),['h','e','l','l','o','\n','w','o','r','l','d'])
    })
    section("segls", function ()
    {
        compare(kseg.segls(hello),[['h','e','l','l','o'],['w','o','r','l','d']])
        compare(kseg.segls(kseg.segls(hello)),[['h','e','l','l','o'],['w','o','r','l','d']])
        compare(kseg.segls(`aaa

bbb`),[['a','a','a'],[],['b','b','b']])
    })
    section("str", function ()
    {
        compare(kseg.str(kseg('hello')),'hello')
        compare(kseg.str(kseg('hello\world')),'hello\world')
        compare(kseg.str(kseg.segls(hello)),hello)
    })
    section("join", function ()
    {
        compare(kseg.join('hello','!','?'),kseg('hello!?'))
        compare(kseg.join(kseg('hello'),'!'),kseg('hello!'))
        compare(kseg.join('hello',kseg('!')),kseg('hello!'))
        compare(kseg.join(kseg.segls('hello'),kseg.segls('world')),kseg.segls("hello\nworld"))
        compare(kseg.join(kseg.segls('hello'),kseg('world')),kseg.segls('hello').concat(kseg('world')))
    })
    section("lines", function ()
    {
        compare(kseg.lines(''),{lines:[''],segls:[[]]})
        compare(kseg.lines(' '),{lines:[' '],segls:[[' ']]})
        compare(kseg.lines('a'),{lines:['a'],segls:[['a']]})
        compare(kseg.lines(hello),{lines:kstr.lines(hello),segls:[kseg('hello'),kseg('world')]})
    })
    section("detab", function ()
    {
        compare(kseg.detab(kseg("ta\tb\tb\ty")),kseg("ta  b   b   y"))
    })
    section("chunks", function ()
    {
        compare(kseg.chunks('a'),[{chunk:'a',index:0,segl:['a']}])
        compare(kseg.chunks(' a'),[{chunk:'a',index:1,segl:['a']}])
        compare(kseg.chunks('ab cd'),[{chunk:'ab',index:0,segl:['a','b']},{chunk:'cd',index:3,segl:['c','d']}])
        compare(kseg.chunks('ab\ncd'),[{chunk:'ab',index:0,segl:['a','b']},{chunk:'cd',index:3,segl:['c','d']}])
        compare(kseg.chunks(kseg('a')),[{chunk:'a',index:0,segl:['a']}])
        compare(kseg.chunks(kseg(' a')),[{chunk:'a',index:1,segl:['a']}])
        compare(kseg.chunks(kseg('ab cd')),[{chunk:'ab',index:0,segl:['a','b']},{chunk:'cd',index:3,segl:['c','d']}])
        compare(kseg.chunks(kseg('ab\ncd')),[{chunk:'ab',index:0,segl:['a','b']},{chunk:'cd',index:3,segl:['c','d']}])
        compare(kseg.chunks(kseg.segls('a')),[{chunk:'a',index:0,segl:['a']}])
        compare(kseg.chunks(kseg.segls(' a')),[{chunk:'a',index:1,segl:['a']}])
        compare(kseg.chunks(kseg.segls('ab cd')),[{chunk:'ab',index:0,segl:['a','b']},{chunk:'cd',index:3,segl:['c','d']}])
        compare(kseg.chunks(kseg.segls('ab\ncd')),[{chunk:'ab',index:0,segl:['a','b']},{chunk:'cd',index:3,segl:['c','d']}])
    })
    section("words", function ()
    {
        compare(kseg.words('a'),[{word:'a',index:0,segl:['a']}])
        compare(kseg.words('ab cd'),[{word:'ab',index:0,segl:['a','b']},{word:'cd',index:3,segl:['c','d']}])
        compare(kseg.words('@b.cd'),[{word:'b',index:1,segl:['b']},{word:'cd',index:3,segl:['c','d']}])
    })
    section("startsWith", function ()
    {
        compare(kseg.startsWith(kseg('something'),''),false)
        compare(kseg.startsWith(kseg('something'),' '),false)
        compare(kseg.startsWith(kseg('something'),'Some'),false)
        compare(kseg.startsWith(kseg('something'),'son'),false)
        compare(kseg.startsWith(kseg('something'),'somethinga'),false)
        compare(kseg.startsWith(kseg('something'),'some'),true)
        compare(kseg.startsWith(kseg('something'),'something'),true)
        compare(kseg.startsWith(kseg('something'),kseg('something')),true)
    })
    section("endsWith", function ()
    {
        compare(kseg.endsWith(kseg('something'),''),false)
        compare(kseg.endsWith(kseg('something'),' '),false)
        compare(kseg.endsWith(kseg('something'),'asomething'),false)
        compare(kseg.endsWith(kseg('something'),'Thing'),false)
        compare(kseg.endsWith(kseg('something'),'thing'),true)
        compare(kseg.endsWith(kseg('something'),'something'),true)
        compare(kseg.endsWith(kseg('something'),kseg('something')),true)
    })
    section("splitAtIndent", function ()
    {
        compare(kseg.splitAtIndent(kseg('something')),[kseg(''),kseg('something')])
        compare(kseg.splitAtIndent(kseg('  some')),[kseg('  '),kseg('some')])
        compare(kseg.splitAtIndent(kseg('    more')),[kseg('    '),kseg('more')])
    })
    section("repeat", function ()
    {
        compare(kseg.repeat(4),kseg("    "))
        compare(kseg.repeat(4,'*'),kseg("****"))
        compare(kseg.repeat(2,'◂▸'),kseg("◂▸◂▸"))
    })
    section("width", function ()
    {
        compare(kseg.width(null),0)
        compare(kseg.width(undefined),0)
        compare(kseg.width(''),0)
        compare(kseg.width([]),0)
        compare(kseg.width({}),0)
        compare(kseg.width('┏━'),2)
        compare(kseg.width(' '),1)
        compare(kseg.width('a'),1)
        compare(kseg.width('💀'),2)
        compare(kseg.width('🔥'),2)
        compare(kseg.width('💩'),2)
        compare(kseg.width('🔨'),2)
        compare(kseg.width('🔧'),2)
        compare(kseg.width('🔨🔧'),4)
        compare(kseg.width('字'),2)
        compare(kseg.width('的'),2)
        compare(kseg.width('块'),2)
        compare(kseg.width('模'),2)
        compare(kseg.width('字的模块'),8)
        compare(kseg.width('🌾'),2)
        compare(kseg.width('👁'),1)
        compare(kseg.width('🖌'),1)
        compare(kseg.width('🖍'),1)
        compare(kseg.width('🛠'),1)
        compare(kseg.width('🧑‍🌾'),2)
        compare(kseg.width('🧑'),2)
        compare(kseg.width('🌞'),2)
        compare(kseg.width('🌙'),2)
        compare(kseg.width('🌿'),2)
        compare(kseg.width('🐦'),2)
        compare(kseg.width('🌊'),2)
        compare(kseg.width('🏰'),2)
        compare(kseg.width('🔮'),2)
        compare(kseg.width('👾'),2)
        compare(kseg.width('🚀'),2)
        compare(kseg.width('🤖'),2)
        compare(kseg.width('🍳'),2)
        compare(kseg.width('🎲'),2)
        compare(kseg.width('🎯'),2)
        compare(kseg.width('🌌'),2)
        compare(kseg.width('🎩'),2)
        compare(kseg.width('📡'),2)
        compare(kseg.width('👽'),2)
        compare(kseg.width('🔭'),2)
        compare(kseg.width('👋'),2)
        compare(kseg.width(kseg("hello world")),11)
    })
    section("trim", function ()
    {
        compare(kseg.trim(' '),kseg(''))
        compare(kseg.trim(' x '),kseg('x'))
        compare(kseg.trim('🖍'),kseg('🖍'))
        compare(kseg.trim('字的模块'),kseg('字的模块'))
        compare(kseg.trim('🧑‍🌾'),kseg('🧑‍🌾'))
        compare(kseg.trim('  🖍    '),kseg('🖍'))
        compare(kseg.trim('  字的模块    '),kseg('字的模块'))
        compare(kseg.trim('  🧑‍🌾   '),kseg('🧑‍🌾'))
    })
    section("headCount", function ()
    {
        compare(kseg.headCount(' x ',' '),1)
        compare(kseg.headCount('***xx***','*'),3)
        compare(kseg.headCount(kseg('***xx***'),'*'),3)
        compare(kseg.headCountWord(' x '),0)
        compare(kseg.headCountWord('xxx  '),3)
        compare(kseg.headCountWord('🧑‍🌾'),0)
        compare(kseg.headCountWord('xx🧑‍🌾'),2)
    })
    section("tailCount", function ()
    {
        compare(kseg.tailCount(' x ',' '),1)
        compare(kseg.tailCount('***xx***','*'),3)
        compare(kseg.tailCount(kseg('***xx***'),'*'),3)
        compare(kseg.tailCountWord(' x '),0)
        compare(kseg.tailCountWord(' xxx'),3)
        compare(kseg.tailCountWord('🧑‍🌾'),0)
        compare(kseg.tailCountWord('🧑‍🌾xx'),2)
    })
    section("numIndent", function ()
    {
        compare(kseg.numIndent('a'),0)
        compare(kseg.numIndent(kseg('a')),0)
        compare(kseg.numIndent('   b'),3)
        compare(kseg.numIndent(kseg('   ')),3)
    })
    section("indexAtWidth", function ()
    {
        compare(kseg.indexAtWidth(kseg(''),0),0)
        compare(kseg.indexAtWidth(kseg(''),1),0)
        compare(kseg.indexAtWidth(kseg('a'),0),0)
        compare(kseg.indexAtWidth(kseg('a'),1),1)
        compare(kseg.indexAtWidth(kseg('a'),2),1)
        compare(kseg.indexAtWidth(kseg('abc'),2),2)
        compare(kseg.indexAtWidth(kseg('ab3'),3),3)
        compare(kseg.indexAtWidth(kseg('ab3'),4),3)
        compare(kseg.indexAtWidth(kseg('🧑🧑'),0),0)
        compare(kseg.indexAtWidth(kseg('🧑🧑'),1),1)
        compare(kseg.indexAtWidth(kseg('🧑🧑'),2),1)
        compare(kseg.indexAtWidth(kseg('🧑🧑'),3),2)
        compare(kseg.indexAtWidth(kseg('🧑🧑'),4),2)
        compare(kseg.indexAtWidth(kseg('🧑🧑'),5),2)
        compare(kseg.indexAtWidth(kseg('🧑🧑'),6),2)
        compare(kseg.indexAtWidth(kseg('🧑‍🌾🧑‍🌾'),0),0)
        compare(kseg.indexAtWidth(kseg('🧑‍🌾🧑‍🌾'),1),1)
        compare(kseg.indexAtWidth(kseg('🧑‍🌾🧑‍🌾'),2),1)
        compare(kseg.indexAtWidth(kseg('🧑‍🌾🧑‍🌾'),3),2)
    })
    section("segiAtWidth", function ()
    {
        compare(kseg.segiAtWidth(kseg(''),0),0)
        compare(kseg.segiAtWidth(kseg(''),1),0)
        compare(kseg.segiAtWidth(kseg('a'),0),0)
        compare(kseg.segiAtWidth(kseg('a'),1),1)
        compare(kseg.segiAtWidth(kseg('a'),2),1)
        compare(kseg.segiAtWidth(kseg('abc'),2),2)
        compare(kseg.segiAtWidth(kseg('ab3'),3),3)
        compare(kseg.segiAtWidth(kseg('ab3'),4),3)
        compare(kseg.segiAtWidth(kseg('🧑🧑'),0),0)
        compare(kseg.segiAtWidth(kseg('🧑🧑'),1),0)
        compare(kseg.segiAtWidth(kseg('🧑🧑'),2),1)
        compare(kseg.segiAtWidth(kseg('🧑🧑'),3),1)
        compare(kseg.segiAtWidth(kseg('🧑🧑'),4),2)
        compare(kseg.segiAtWidth(kseg('🧑🧑'),5),2)
        compare(kseg.segiAtWidth(kseg('🧑🧑'),6),2)
    })
    section("spanForClosestWordAtColumn", function ()
    {
        compare(kseg.spanForClosestWordAtColumn(kseg('abc def'),0),[0,3])
        compare(kseg.spanForClosestWordAtColumn(kseg('abc def'),1),[0,3])
        compare(kseg.spanForClosestWordAtColumn(kseg('abc def'),2),[0,3])
        compare(kseg.spanForClosestWordAtColumn(kseg('abc def'),3),[0,3])
        compare(kseg.spanForClosestWordAtColumn(kseg('abc def'),4),[4,7])
        compare(kseg.spanForClosestWordAtColumn(kseg('ab  def'),3),[0,2])
        compare(kseg.spanForClosestWordAtColumn(kseg('ab  def'),4),[4,7])
        compare(kseg.spanForClosestWordAtColumn(kseg('     '),0),[0,0])
        compare(kseg.spanForClosestWordAtColumn(kseg('     '),2),[2,2])
        compare(kseg.spanForClosestWordAtColumn(kseg('     '),3),[3,3])
        compare(kseg.spanForClosestWordAtColumn(kseg('   xy'),2),[3,5])
        compare(kseg.spanForClosestWordAtColumn(kseg('xy   '),3),[0,2])
    })
    section("collectGraphemes", function ()
    {
        compare(kseg.collectGraphemes(kseg('hello world')),['h','e','l','o',' ','w','r','d'])
        text = kseg(` 0000000  000       0000000    0000000  00000000   0000000  000000000  000   000   0000000   00000000   0000000    
000       000      000   000  000       000       000          000     000 0 000  000   000  000   000  000   000  
000       000      000   000  0000000   0000000   0000000      000     000000000  000   000  0000000    000   000  
000       000      000   000       000  000            000     000     000   000  000   000  000   000  000   000  
 0000000  0000000   0000000   0000000   00000000  0000000      000     00     00   0000000   000   000  0000000    `)
        compare(kseg.collectGraphemes(text),[' ','0','\n'])
    })
}
toExport["kseg"]._section_ = true
toExport._test_ = true
export default toExport

File: kakao.app/kao/lua/kxk/test/kseg.lua
--[[
    ███   ███   ███████  ████████   ███████       █████████  ████████   ███████  █████████  
    ███  ███   ███       ███       ███               ███     ███       ███          ███     
    ███████    ███████   ███████   ███  ████         ███     ███████   ███████      ███     
    ███  ███        ███  ███       ███   ███         ███     ███            ███     ███     
    ███   ███  ███████   ████████   ███████          ███     ████████  ███████      ███     
--]]

kxk = require "kxk/kxk"

test("kseg", function()
    test("static", function()
        test.cmp(kseg.rep(4):str(), "    ")
        test.cmp(kseg.rep(4, 'x'):str(), "xxxx")
        test.cmp(kseg.rep(2, 'xy'):str(), "xyxy")
        
        test.cmp(kseg.segiAtWidth(kseg(''), 0), 1)
        test.cmp(kseg.segiAtWidth(kseg(''), 1), 1)
        test.cmp(kseg.segiAtWidth(kseg('a'), 1), 1)
        test.cmp(kseg.segiAtWidth(kseg('a'), 2), 2)
        test.cmp(kseg.segiAtWidth(kseg('abc'), 2), 2)
        test.cmp(kseg.segiAtWidth(kseg('ab3'), 3), 3)
        test.cmp(kseg.segiAtWidth(kseg('ab3'), 4), 4)
        test.cmp(kseg.segiAtWidth(kseg('ab3'), 5), 4)
        test.cmp(kseg.segiAtWidth(kseg('ab3'), 6), 4)
        
        test.cmp(kseg.segiAtWidth(kseg('🧑🧑'), 0), 1)
        test.cmp(kseg.segiAtWidth(kseg('🧑🧑'), 1), 1)
        test.cmp(kseg.segiAtWidth(kseg('🧑🧑'), 2), 1)
        test.cmp(kseg.segiAtWidth(kseg('🧑🧑'), 3), 2)
        test.cmp(kseg.segiAtWidth(kseg('🧑🧑'), 4), 2)
        test.cmp(kseg.segiAtWidth(kseg('🧑🧑'), 5), 3)
        test.cmp(kseg.segiAtWidth(kseg('🧑🧑'), 6), 3)
    end)
    
    test("segls", function()
        local lines = kstr.lines [[
123
456

abc
def]]
        
        test.cmp(lines, array("123", "456", "", "abc", "def"))
        
        local segls = kseg.segls([[
323
456

abc
def]])
        
        test.cmp(segls, array(kseg("323"), kseg("456"), kseg(""), kseg("abc"), kseg("def")))
        
        test.cmp(kseg.str(segls), [[
323
456

abc
def]])
        
        test.cmp(kseg.segls([[
◆1
◆2
◆3
◆4]], "\n"), array(kseg("◆1"), kseg("◆2"), kseg("◆3"), kseg("◆4")))
    end)
    
    test("construct", function()
        local s = kseg()
        test.cmp(s:str(), "")
        
        s = kseg("  ")
        test.cmp(s:str(), "  ")
        
        s = kseg(kseg("xx"))
        test.cmp(s:str(), "xx")
        
        s = kseg(kseg(kseg("yy")))
        test.cmp(s:str(), "yy")
        
        local st = strg("zz")
        s = kseg(strg("zz"))
        test.cmp(s:str(), "zz")
        
        s = kseg("abc")
        local z = kseg(s)
        test.cmp(s:str(), "abc")
        test.cmp(z:str(), "abc")
        test.cmp((s == z), false)
        z:push("d")
        test.cmp(s:str(), "abc")
        test.cmp(z:str(), "abcd")
    end)
    
    test("startsWith", function()
        local s = kseg("a  b")
        test.cmp(s:startsWith("a"), true)
        test.cmp(s:startsWith("a "), true)
        test.cmp(s:startsWith("a  b"), true)
        test.cmp(s:startsWith("a  bc"), false)
        test.cmp(s:startsWith("b"), false)
        
        s = kseg("# com")
        test.cmp(s:startsWith(kseg("#")), true)
        test.cmp(s:startsWith(kseg("# ")), true)
        test.cmp(s:startsWith(kseg("# c")), true)
        test.cmp(s:startsWith(kseg(" ")), false)
        test.cmp(s:startsWith(kseg("c")), false)
        
        test.cmp(kseg.startsWith(s, kseg("#")), true)
    end)
    
    test("endsWith", function()
        local s = kseg("a  b")
        test.cmp(s:endsWith("b"), true)
        test.cmp(s:endsWith(" b"), true)
        test.cmp(s:endsWith("a  b"), true)
        test.cmp(s:endsWith(" a  b"), false)
        test.cmp(s:endsWith("a"), false)
    end)
    
    test("find", function()
        local s = kseg("a  b")
        test.cmp(s:find(" "), 2)
        test.cmp(s:find("  "), 2)
        test.cmp(s:find(" b"), 3)
        test.cmp(s:find("b"), 4)
    end)
    
    test("pop", function()
        local s = kseg("../")
        s:pop()
        test.cmp(s:str(), "..")
        
        s = kseg("◂▸○●")
        s:pop()
        test.cmp(s:str(), "◂▸○")
        s:pop()
        test.cmp(s:str(), "◂▸")
    end)
    
    test("push", function()
        local s = kseg()
        test.cmp(#s, 0)
        s:push("●")
        test.cmp(s:str(), "●")
        test.cmp(#s, 1)
        s:push("▪")
        test.cmp(s:str(), "●▪")
        test.cmp(#s, 2)
        s:push(" ")
        test.cmp(s:str(), "●▪ ")
        test.cmp(#s, 3)
    end)
    
    test("rpad", function()
        local s = kseg("▴")
        s:rpad(2)
        test.cmp(s:str(), "▴ ")
        s:rpad(4)
        test.cmp(s:str(), "▴   ")
        s:rpad(2)
        test.cmp(s:str(), "▴   ")
    end)
    
    test("trim", function()
        local s = kseg(" ▴  ")
        s:trim()
        test.cmp(s:str(), "▴")
        
        s = kseg(" ▴  ")
        s:rtrim()
        test.cmp(s:str(), " ▴")
        
        s = kseg(" ▴  ")
        s:ltrim()
        test.cmp(s:str(), "▴  ")
        
        s = kseg("\n\n ▴  \n  \n")
        s:trim()
        test.cmp(s:str(), "▴")
    end)
    
    test("slice", function()
        local s = kseg("abc")
        test.cmp(s:slice(1):str(), "abc")
        test.cmp(s:slice(2):str(), "bc")
        test.cmp(s:slice(3):str(), "c")
        test.cmp(s:slice(4, 3):str(), "")
        test.cmp(s:slice(4, 4):str(), "")
        test.cmp(s:slice(4):str(), "")
    end)
    
    test("emojii", function()
        local s = kseg("🧑")
        test.cmp(#s, 1)
        
        s = kseg("🧑‍🌾")
        test.cmp(#s, 3)
        test.cmp(s[#s], "🌾")
        s:pop()
        s:pop()
        test.cmp(#s, 1)
        test.cmp(s:str(), "🧑")
    end)
    
    test("segiAtWidth", function()
        test.cmp(kseg.segiAtWidth(kseg("abc"), 1), 1)
        test.cmp(kseg.segiAtWidth(kseg("abc"), 2), 2)
        test.cmp(kseg.segiAtWidth(kseg("abc"), 3), 3)
        test.cmp(kseg.segiAtWidth(kseg("abc"), 4), 4)
        
        test.cmp(kseg.segiAtWidth(kseg("▸◌◂"), 1), 1)
        test.cmp(kseg.segiAtWidth(kseg("▸◌◂"), 2), 2)
        test.cmp(kseg.segiAtWidth(kseg("▸◌◂"), 3), 3)
        test.cmp(kseg.segiAtWidth(kseg("▸◌◂"), 4), 4)
        
        test.cmp(kseg.segiAtWidth("abc", 1), 1)
        test.cmp(kseg.segiAtWidth("abc", 2), 2)
        test.cmp(kseg.segiAtWidth("abc", 3), 3)
        test.cmp(kseg.segiAtWidth("abc", 4), 4)
        
        test.cmp(kseg.segiAtWidth("▸◌◂", 1), 1)
        test.cmp(kseg.segiAtWidth("▸◌◂", 2), 2)
        test.cmp(kseg.segiAtWidth("▸◌◂", 3), 3)
        test.cmp(kseg.segiAtWidth("▸◌◂", 4), 4)
        
        test.cmp(kseg.segiAtWidth(kseg('🧑🧑'), 1), 1)
        test.cmp(kseg.segiAtWidth(kseg('🧑🧑'), 2), 1)
        test.cmp(kseg.segiAtWidth(kseg('🧑🧑'), 3), 2)
        test.cmp(kseg.segiAtWidth(kseg('🧑🧑'), 4), 2)
        test.cmp(kseg.segiAtWidth(kseg('🧑🧑'), 5), 3)
        test.cmp(kseg.segiAtWidth(kseg('🧑🧑'), 6), 3)
    end)
    
    test("indexAtWidth", function()
        test.cmp(kseg.indexAtWidth(kseg(''), 1), 1)
        test.cmp(kseg.indexAtWidth(kseg('a'), 1), 1)
        test.cmp(kseg.indexAtWidth(kseg('a'), 2), 2)
        test.cmp(kseg.indexAtWidth(kseg('abc'), 2), 2)
        test.cmp(kseg.indexAtWidth(kseg('ab3'), 3), 3)
        test.cmp(kseg.indexAtWidth(kseg('ab3'), 4), 4)
        
        test.cmp(kseg.indexAtWidth(kseg('🧑🧑'), 1), 1)
        test.cmp(kseg.indexAtWidth(kseg('🧑🧑'), 2), 2)
        test.cmp(kseg.indexAtWidth(kseg('🧑🧑'), 3), 2)
        test.cmp(kseg.indexAtWidth(kseg('🧑🧑'), 4), 3)
        test.cmp(kseg.indexAtWidth(kseg('🧑🧑'), 5), 3)
        test.cmp(kseg.indexAtWidth(kseg('🧑🧑'), 6), 3)
        
        -- kseg.indexAtWidth kseg('🧑‍🌾🧑‍🌾')  1 ▸ 1
        -- kseg.indexAtWidth kseg('🧑‍🌾🧑‍🌾')  2 ▸ 2
        -- kseg.indexAtWidth kseg('🧑‍🌾🧑‍🌾')  3 ▸ 2
        -- kseg.indexAtWidth kseg('🧑‍🌾🧑‍🌾')  4 ▸ 3
    end)
    
    test("eql", function()
        test.cmp((kseg("abc") == kseg("abc")), false)
        
        test.cmp(kseg("abc"):eql(kseg("abc")), true)
        
        test.cmp(kseg.eql("abc", "abc"), true)
        test.cmp(kseg.eql("abc", kseg("abc")), true)
        test.cmp(kseg.eql(kseg("abc"), "abc"), true)
    end)
    
    test("is", function()
        local a = kseg("")
        test.cmp(a:is(kseg), true)
        test.cmp(a:is(array), true)
        test.cmp(a:is(table), false)
        test.cmp(a:is(false), false)
        test.cmp(a:is(true), false)
        test.cmp(a:is(nil), false)
    end)
    
    test("count", function()
        local s = kseg("__hellooo")
        test.cmp((s[#s] == "o"), true)
        test.cmp((s:slice(#s):str() == "o"), true)
        test.cmp(kseg.tailCount(s, "o"), 3)
        test.cmp(kseg.tailCount("__hellooo", "o"), 3)
        
        test.cmp(kseg.headCount(s, "_"), 2)
        test.cmp(kseg.headCount("__hellooo", "_"), 2)
        
        test.cmp(kseg.tailCount("◂◂○○●", "●"), 1)
        test.cmp(kseg.headCount("◂◂○○●", "◂"), 2)
        
        test.cmp(kseg.tailCountWord(kseg("1  2 33")), 2)
        test.cmp(kseg.tailCountWord(kseg("x233 ")), 0)
        
        test.cmp(kseg.headCountWord(kseg("1  2 33")), 1)
        test.cmp(kseg.headCountWord(kseg("  2 33")), 0)
        test.cmp(kseg.headCountWord(kseg(".  2 33")), 0)
        test.cmp(kseg.headCountWord(kseg("x233")), 4)
        test.cmp(kseg.headCountWord(kseg("x233 ")), 4)
        
        test.cmp(kseg.tailCount(" ", ""), 0)
        test.cmp(kseg.tailCount(" ", " "), 1)
        test.cmp(kseg.tailCount("  ", " "), 2)
        
        test.cmp(kseg.tailCountChunk(kseg("a bc")), 2)
        test.cmp(kseg.tailCountChunk(kseg("a b c")), 1)
        test.cmp(kseg.tailCountChunk(kseg("a b c  ")), 0)
        test.cmp(kseg.tailCountChunk(kseg("abc  ")), 0)
        test.cmp(kseg.tailCountChunk(kseg("  ")), 0)
        
        test.cmp(kseg.headCountChunk(kseg("a bc")), 1)
        test.cmp(kseg.headCountChunk(kseg("ab c")), 2)
        test.cmp(kseg.headCountChunk(kseg(" a b c  ")), 0)
        test.cmp(kseg.headCountChunk(kseg("abc  ")), 3)
        test.cmp(kseg.headCountChunk(kseg("  ")), 0)
    end)
    
    test("spanForClosestWordAtColumn", function()
        test.cmp(kseg.spanForClosestWordAtColumn(kseg('abc def'), 1), array(1, 4))
        test.cmp(kseg.spanForClosestWordAtColumn(kseg('abc def'), 2), array(1, 4))
        test.cmp(kseg.spanForClosestWordAtColumn(kseg('abc def'), 3), array(1, 4))
        test.cmp(kseg.spanForClosestWordAtColumn(kseg('abc def'), 4), array(5, 8))
        test.cmp(kseg.spanForClosestWordAtColumn(kseg('ab  def'), 3), array(1, 3))
        test.cmp(kseg.spanForClosestWordAtColumn(kseg('ab  def'), 4), array(5, 8))
        
        test.cmp(kseg.spanForClosestWordAtColumn(kseg('     '), 0), array(0, 0))
        test.cmp(kseg.spanForClosestWordAtColumn(kseg('     '), 2), array(2, 2))
        test.cmp(kseg.spanForClosestWordAtColumn(kseg('     '), 3), array(3, 3))
        
        test.cmp(kseg.spanForClosestWordAtColumn(kseg('   xy'), 2), array(4, 6))
        test.cmp(kseg.spanForClosestWordAtColumn(kseg('xy   '), 3), array(1, 3))
        
        test.cmp(kseg.spanForClosestWordAtColumn("   aa123   ", 1), array(4, 9))
    end)
    end)